### PR TITLE
Spatial: Translate GeometryType and OgcGeometryType

### LINKED ISSUE #13280: Sqlite Spatial: Translate GeometryType
Unfortunately, the result of GeometryType is different between NTS and SpatiaLite:

Lib | Result
--- | ---
NTS | Point
SpatiaLite | POINT Z

We really need some way of determining an instance type to enable efficiently querying heterogeneous collections.

We could either:
* Return the value directly--require handling case and dimension in the application; values change between server and client eval.
* Massage the strings to look like NTS
* Don't translate--we should provide an alternative like translating `OgcGeometryType` using a big `CASE` expression

### PR DESCRIPTION
**Pull request by bricelam**
Fixes #13280

**Providers**, this adds a class for generating SQL CASE expressions.

### CODE REVIEW

File: src/EFCore.Relational/Query/Expressions/CaseExpression.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        /// </summary>
+        /// <value> One of the <see cref="ExpressionType"/> values. </value>
+        public override ExpressionType NodeType
+            => ExpressionType.Extension;
+
+        /// <summary>
+        ///     Gets the static type of the expression that this <see cref="Expression"/> represents.
+        /// </summary>
+        /// <value> The <see cref="Type"/> that represents the static type of the expression. </value>
+        public override Type Type
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/CaseExpression.cs]: Do we have a check for Empty WhenThenList?

File: src/EFCore.Relational/Query/Expressions/CaseExpression.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        /// <returns> The expression being visited, or an expression which should replace it in the tree. </returns>
+        protected override Expression VisitChildren(ExpressionVisitor visitor)
+        {
+            var newOperand = visitor.Visit(Operand);
+
+            var whenThenListChanged = false;
+            var newWhenThenList = new List<WhenThen>();
+            foreach (var whenThen in WhenThenList)
+            {
+                var newTest = visitor.Visit(whenThen.Test);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Expressions/CaseExpression.cs]: Bit counter intuitive to have `WhenThen` but not expression type. If it is expression type then this logic can be moved to itself.

File: src/EFCore.Relational/Query/Expressions/CaseExpression.cs (hunk 1 above)
Comment (bricelam) [reply src/EFCore.Relational/Query/Expressions/CaseExpression.cs]: Will validate types too.

**Comment by bricelam** [comment]
Updated. Added type validation, and renamed `WhenThen` to `CaseWhenClause`.

File: src/EFCore.Relational/Query/Expressions/CaseWhenClause.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            Check.NotNull(result, nameof(result));
+
+            Test = test;
+            Result = result;
+        }
+
+        /// <summary>
+        ///     Gets the when operand expression.
+        /// </summary>
+        public virtual Expression Test { get; }
```
Comment (bricelam) [inline src/EFCore.Relational/Query/Expressions/CaseWhenClause.cs]: Fun fact: The SQL spec actually allows for multiple tests, but I don't know of a database that implements it.

``` sql
CASE Name
    WHEN 'Brice', 'BRICE', 'brice' THEN 'me'
END
```

File: src/EFCore.Relational/Query/Expressions/CaseExpression.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+using Microsoft.EntityFrameworkCore.Internal;
+using Microsoft.EntityFrameworkCore.Query.Sql;
+using Microsoft.EntityFrameworkCore.Utilities;
+
+namespace Microsoft.EntityFrameworkCore.Query.Expressions
+{
+    /// <summary>
+    ///     Represents a SQL CASE expression.
+    /// </summary>
+    public class CaseExpression : Expression
```
Comment (maumar) [inline src/EFCore.Relational/Query/Expressions/CaseExpression.cs]: also add IPrintable, for the pretty-printing hook

File: src/EFCore.Relational/Properties/RelationalStrings.resx [hunk 5]
```diff
@@ -483,4 +483,13 @@
   <data name="RelationalCloneNotImplemented" xml:space="preserve">
     <value>The '{mapping}' does not support 2.2 style type mapping. The database provider needs to be updated to support the full set of mapping customization.</value>
   </data>
+  <data name="CaseElseResultTypeUnexpected" xml:space="preserve">
+    <value>An else result type of '{elseResultType}' is invalid. The expected type is '{resultType}'.</value>
+  </data>
+  <data name="CaseWhenClauseResultTypeUnexpected" xml:space="preserve">
+    <value>A when result type of '{whenResultType}' is invalid. The expected type is '{resultType}'.</value>
```
[the remark this answers is not in the document]
Comment (bricelam) [reply src/EFCore.Relational/Properties/RelationalStrings.resx]: Will reword. Maybe something like "The result type 'blah' of a when clause..."

**Pull request merged**